/memberships returns status: 400 (code: 9106) — how should wrangler react?

### What versions & operating system are you using?

wrangler 4.84.1
macos latest

### Please provide a link to a minimal reproduction

see below

### Describe the Bug

I see this and think something expired / invalid about my session:

```
➜ npx wrangler d1 list

 ⛅️ wrangler 4.84.1
───────────────────

✘ [ERROR] A request to the Cloudflare API (/memberships) failed.

  Authentication failed (status: 400) [code: 9106]

  If you think this is a bug, please open an issue at:
  https://github.com/cloudflare/workers-sdk/issues/new/choose
```

Here's the 4xx error:

```json
{
  "code": 9106,
  "message": "Missing X-Auth-Key, X-Auth-Email or Authorization headers"
}
```

Then I run `npx wrangler login` and _only then_ realize that wrangler is using and API token I set in my env:

```
 ⛅️ wrangler 4.84.1
───────────────────
✘ [ERROR] You are logged in with an API Token. Unset the CLOUDFLARE_API_TOKEN in the environment to log in via OAuth.
```

I think I get why this happens (because API tokens don't currently support FGA, which I have enabled on my account? slash - work that we are in progress on for the Workers API)

But thinking more broadly about how we handle 4xx errors — should we handle more gracefully?

- suggest that the user `npx wrangler logout` / `npx wrangler login`?
- expose earlier that an API token is being used by wrangler? and where that is coming from?
- attempt this for them?
- something else?

Thinking broadly about — how do we help humans and agents figure out what to do here, because won't be only time when there is 4xx

### Please provide any relevant error logs

_No response_

--- a/.changeset/fix-auth-error-hint-env-var.md
+++ b/.changeset/fix-auth-error-hint-env-var.md
@@ -1,0 +1,7 @@
+---
+"wrangler": patch
+---
+
+fix: show actionable hint when `/memberships` returns a bad-credentials error (code 9106)
+
+Previously, `wrangler` threw a raw Cloudflare API error ("Missing X-Auth-Key, X-Auth-Email or Authorization headers") with no guidance. Now it emits a `UserError` explaining that an environment variable such as `CLOUDFLARE_API_TOKEN`, `CLOUDFLARE_API_KEY`, or `CLOUDFLARE_EMAIL` may be set to an invalid value, and suggests running `wrangler logout` / `wrangler login` to re-authenticate.

--- a/packages/wrangler/src/__tests__/kv/key.test.ts
+++ b/packages/wrangler/src/__tests__/kv/key.test.ts
@@ -1246,7 +1246,9 @@ describe("kv", () => {
 						runWrangler("kv key get --remote key --namespace-id=xxxx")
 					).rejects.toThrowErrorMatchingInlineSnapshot(`
 						[Error: Failed to automatically retrieve account IDs for the logged in user.
-						You may have incorrect permissions on your API token. You can skip this account check by adding an \`account_id\` in your Wrangler configuration file, or by setting the value of CLOUDFLARE_ACCOUNT_ID]
+						You may have incorrect permissions on your API token, or an environment variable such as CLOUDFLARE_API_TOKEN, CLOUDFLARE_API_KEY, or CLOUDFLARE_EMAIL may be set to an invalid value.
+						Check your environment and unset or correct any Cloudflare credential variables, or run \`wrangler logout\` followed by \`wrangler login\` to re-authenticate.
+						You can also skip this account check by adding an \`account_id\` in your Wrangler configuration file, or by setting the value of CLOUDFLARE_ACCOUNT_ID]
 					`);
 				});
 

--- a/packages/wrangler/src/__tests__/user.test.ts
+++ b/packages/wrangler/src/__tests__/user.test.ts
@@ -1014,6 +1014,36 @@ describe("User", () => {
 			);
 		});
 
+		it("should include env-var hint when /memberships returns 9106 and /accounts is empty", async ({
+			expect,
+		}) => {
+			msw.use(
+				http.get("*/accounts", () => HttpResponse.json(createFetchResult([])), {
+					once: true,
+				}),
+				http.get(
+					"*/memberships",
+					() => {
+						return HttpResponse.json({
+							success: false,
+							errors: [
+								{
+									code: 9106,
+									message: "Authentication failed",
+								},
+							],
+							result: null,
+						});
+					},
+					{ once: true }
+				)
+			);
+
+			await expect(fetchAllAccounts({})).rejects.toThrowError(
+				/CLOUDFLARE_API_TOKEN/
+			);
+		});
+
 		it("should propagate /memberships errors that are not 9106 or 10000", async ({
 			expect,
 		}) => {

--- a/packages/wrangler/src/user/fetch-accounts.ts
+++ b/packages/wrangler/src/user/fetch-accounts.ts
@@ -17,13 +17,13 @@ import type { ComplianceConfig } from "@cloudflare/workers-utils";
 //     auth (e.g. tokens missing the membership read scope).
 const MEMBERSHIPS_INACCESSIBLE_CODES = [9106, 10000];
 
-function isMembershipsInaccessible(err: unknown): boolean {
-	const code = (err as { code?: number } | undefined)?.code;
-	return code !== undefined && MEMBERSHIPS_INACCESSIBLE_CODES.includes(code);
+function getErrorCode(err: unknown): number | undefined {
+	return (err as { code?: number } | undefined)?.code;
 }
 
-function getMembershipsErrorCode(err: unknown): number | undefined {
-	return (err as { code?: number } | undefined)?.code;
+function isMembershipsInaccessible(err: unknown): boolean {
+	const code = getErrorCode(err);
+	return code !== undefined && MEMBERSHIPS_INACCESSIBLE_CODES.includes(code);
 }
 
 /**
@@ -78,9 +78,9 @@ export async function fetchAllAccounts(
 				// Fall back to `/accounts`, which is already scoped to what the auth can use.
 				return accountsRes.value;
 			}
-			// 9106 specifically can also be returned when an environment variable like
+			// 9106 specifically can be returned when an environment variable like
 			// CLOUDFLARE_API_TOKEN is set to an invalid value — surface that hint.
-			const errCode = getMembershipsErrorCode(membershipsRes.reason);
+			const errCode = getErrorCode(membershipsRes.reason);
 			if (errCode === 9106) {
 				throw new UserError(
 					`Failed to automatically retrieve account IDs for the logged in user.

--- a/packages/wrangler/src/user/fetch-accounts.ts
+++ b/packages/wrangler/src/user/fetch-accounts.ts
@@ -22,6 +22,10 @@ function isMembershipsInaccessible(err: unknown): boolean {
 	return code !== undefined && MEMBERSHIPS_INACCESSIBLE_CODES.includes(code);
 }
 
+function getMembershipsErrorCode(err: unknown): number | undefined {
+	return (err as { code?: number } | undefined)?.code;
+}
+
 /**
  * Fetches the set of accounts that the current login auth can actually use.
  *
@@ -73,6 +77,18 @@ export async function fetchAllAccounts(
 			if (accountsRes.status === "fulfilled" && accountsRes.value.length > 0) {
 				// Fall back to `/accounts`, which is already scoped to what the auth can use.
 				return accountsRes.value;
+			}
+			// 9106 specifically can also be returned when an environment variable like
+			// CLOUDFLARE_API_TOKEN is set to an invalid value — surface that hint.
+			const errCode = getMembershipsErrorCode(membershipsRes.reason);
+			if (errCode === 9106) {
+				throw new UserError(
+					`Failed to automatically retrieve account IDs for the logged in user.
+You may have incorrect permissions on your API token, or an environment variable such as CLOUDFLARE_API_TOKEN, CLOUDFLARE_API_KEY, or CLOUDFLARE_EMAIL may be set to an invalid value.
+Check your environment and unset or correct any Cloudflare credential variables, or run \`wrangler logout\` followed by \`wrangler login\` to re-authenticate.
+You can also skip this account check by adding an \`account_id\` in your ${configFileName(undefined)} file, or by setting the value of CLOUDFLARE_ACCOUNT_ID`,
+					{ telemetryMessage: "user account fetch permission denied" }
+				);
 			}
 			throw new UserError(
 				`Failed to automatically retrieve account IDs for the logged in user.

